language/bytecode-verifier/src/borrow_graph.rs

```diff
@@ -0,0 +1,245 @@
+                .filter(|x| !x.label.is_empty() && x.label[0] == label[0])
+                .cloned()
+                .collect();
+            let from_edge_set = self.0.get_mut(&from).unwrap();
+            edge_set.iter().for_each(|x| {
+                from_edge_set.remove(x);
+            });
+            from_edge_set.insert(new_edge);
+            edge_set.iter_mut().for_each(|x| {
+                x.label.remove(0);
+            });
+            self.0.insert(to, edge_set.into_iter().collect());
+        }
+    }
+
+    /// removes `nonce` and appropriately concatenates each incoming edge with each outgoing edge of `nonce`
+    pub fn remove_nonce(&mut self, nonce: Nonce) {
+        checked_assume!(self.invariant());
+        let nonce_edge_set = self.0.remove(&nonce).unwrap();
+        let mut removed_edges = BTreeMap::new();
+        for (n, es) in &self.0 {
+            removed_edges.insert(
+                n.clone(),
+                es.iter()
+                    .filter(|x| x.to == nonce)
+                    .cloned()
+                    .collect::<Vec<_>>(),
+            );
+        }
+        for (n, es) in &removed_edges {
+            es.iter().for_each(|removed_edge| {
+                let n_edge_set_ref = self.0.get_mut(n).unwrap();
+                n_edge_set_ref.remove(removed_edge);
+                nonce_edge_set.iter().for_each(|nonce_edge| {
+                    if removed_edge.edge_type == EdgeType::Strong {
+                        let mut new_label = vec![];
+                        new_label.append(&mut removed_edge.label.clone());
+                        new_label.append(&mut nonce_edge.label.clone());
+                        let edge = Edge {
+                            edge_type: nonce_edge.edge_type,
+                            label: new_label,
+                            to: nonce_edge.to,
+                        };
+                        n_edge_set_ref.insert(edge);
+                    } else {
+                        let edge = Edge {
+                            edge_type: EdgeType::Weak,
+                            label: removed_edge.label.clone(),
+                            to: nonce_edge.to,
+                        };
+                        n_edge_set_ref.insert(edge);
+                    }
+                });
+            });
+        }
+        checked_verify!(self.invariant());
+        checked_postcondition!(!self.0.contains_key(&nonce));
+    }
+
+    /// renames nonces in `self` according to `nonce_map`
+    pub fn rename_nonces(&self, nonce_map: BTreeMap<Nonce, Nonce>) -> Self {
+        checked_assume!(self.invariant());
+        let mut new_graph = BTreeMap::new();
+        for (n, es) in &self.0 {
+            new_graph.insert(
+                nonce_map[n],
+                es.iter()
+                    .map(|x| Edge {
+                        edge_type: x.edge_type,
+                        label: x.label.clone(),
+                        to: nonce_map[&x.to],
```

Comment:
Could [struct update syntax](https://doc.rust-lang.org/book/ch05-01-defining-structs.html#creating-instances-from-other-instances-with-struct-update-syntax) be useful here?

Reply:
Let us stick with the explicit syntax for now.